File: server/services/parser.js
const XLSX = require('xlsx');

const REQUIRED_COLS = [
  'Employee ID',
  'Nickname / Name',
  'Cost Centre',
  'Gross Salary',
  'EPF Employer',
  'EIS Employer',
  'SOCSO Employer',
  'HRDF',
  'MTD',
  'CTC Hexa',
  'Net Salary',
];

function toNum(val) {
  if (val === null || val === undefined || val === '') return 0;
  const n = parseFloat(val.toString().replace(/,/g, ''));
  return isNaN(n) ? 0 : n;
}

function parseExcel(filePath) {
  const workbook = XLSX.readFile(filePath, { cellDates: true });
  const entities = [];

  for (const sheetName of workbook.SheetNames) {
    const sheet = workbook.Sheets[sheetName];
    const rows = XLSX.utils.sheet_to_json(sheet, { header: 1, defval: null, raw: false });

    if (!rows || rows.length < 2) continue;

    const headerRow = rows[0];
    const colMap = {};
    headerRow.forEach((cell, idx) => {
      if (cell !== null && cell !== undefined) {
        colMap[cell.toString().trim()] = idx;
      }
    });

    const missingCols = REQUIRED_COLS.filter((c) => !(c in colMap));

    const employees = [];
    for (let i = 1; i < rows.length; i++) {
      const row = rows[i];
      if (!row) continue;

      const employeeId = row[colMap['Employee ID']];
      if (employeeId === null || employeeId === undefined || employeeId.toString().trim() === '') continue;

      const ctcHexa = toNum(row[colMap['CTC Hexa']]);
      if (ctcHexa === 0) continue;

      employees.push({
        employeeId: employeeId.toString().trim(),
        name: (row[colMap['Nickname / Name']] || '').toString().trim(),
        costCentre: (row[colMap['Cost Centre']] || '').toString().trim(),
        grossSalary: toNum(row[colMap['Gross Salary']]),
        epfEmployer: toNum(row[colMap['EPF Employer']]),
        eisEmployer: toNum(row[colMap['EIS Employer']]),
        socsoEmployer: toNum(row[colMap['SOCSO Employer']]),
        hrdf: toNum(row[colMap['HRDF']]),
        mtd: toNum(row[colMap['MTD']]),
        ctcHexa,
        netSalary: toNum(row[colMap['Net Salary']]),
      });
    }

    if (employees.length === 0) continue;

    const totalCTC = employees.reduce((sum, e) => sum + e.ctcHexa, 0);

    entities.push({
      sheetName,
      employees,
      totalCTC: Math.round(totalCTC * 100) / 100,
      missingColumns: missingCols,
    });
  }

  return entities;
}

module.exports = { parseExcel };

File: app/services/parser.py
import io
from openpyxl import load_workbook

REQUIRED_COLS = [
    "Employee ID", "Nickname / Name", "Cost Centre",
    "Gross Salary", "EPF Employer", "EIS Employer",
    "SOCSO Employer", "HRDF", "MTD", "CTC Hexa", "Net Salary",
]


def _to_num(val) -> float:
    if val is None or val == "":
        return 0.0
    try:
        return float(str(val).replace(",", ""))
    except (ValueError, TypeError):
        return 0.0


def parse_excel_buffer(data: bytes) -> list[dict]:
    wb = load_workbook(io.BytesIO(data), read_only=True, data_only=True)
    entities = []

    for sheet_name in wb.sheetnames:
        ws = wb[sheet_name]
        rows = list(ws.iter_rows(values_only=True))
        if not rows or len(rows) < 2:
            continue

        header_row = rows[0]
        col_map: dict[str, int] = {}
        for idx, cell in enumerate(header_row):
            if cell is not None:
                col_map[str(cell).strip()] = idx

        missing_cols = [c for c in REQUIRED_COLS if c not in col_map]

        employees = []
        for row in rows[1:]:
            if not row:
                continue
            emp_id_idx = col_map.get("Employee ID")
            if emp_id_idx is None:
                continue
            emp_id_val = row[emp_id_idx]
            if emp_id_val is None or str(emp_id_val).strip() == "":
                continue

            ctc_hexa = _to_num(row[col_map["CTC Hexa"]] if "CTC Hexa" in col_map else None)
            if ctc_hexa == 0:
                continue

            employees.append({
                "employeeId": str(emp_id_val).strip(),
                "name": str(row[col_map["Nickname / Name"]] or "").strip() if "Nickname / Name" in col_map else "",
                "costCentre": str(row[col_map["Cost Centre"]] or "").strip() if "Cost Centre" in col_map else "",
                "grossSalary": _to_num(row[col_map["Gross Salary"]] if "Gross Salary" in col_map else None),
                "epfEmployer": _to_num(row[col_map["EPF Employer"]] if "EPF Employer" in col_map else None),
                "eisEmployer": _to_num(row[col_map["EIS Employer"]] if "EIS Employer" in col_map else None),
                "socsoEmployer": _to_num(row[col_map["SOCSO Employer"]] if "SOCSO Employer" in col_map else None),
                "hrdf": _to_num(row[col_map["HRDF"]] if "HRDF" in col_map else None),
                "mtd": _to_num(row[col_map["MTD"]] if "MTD" in col_map else None),
                "ctcHexa": ctc_hexa,
                "netSalary": _to_num(row[col_map["Net Salary"]] if "Net Salary" in col_map else None),
            })

        if not employees:
            continue

        total_ctc = round(sum(e["ctcHexa"] for e in employees), 2)
        entities.append({
            "sheetName": sheet_name.strip(),
            "employees": employees,
            "totalCTC": total_ctc,
            "missingColumns": missing_cols,
        })

    wb.close()
    return entities
